compositor_web/src/wasm.rs

```diff
@@ -0,0 +1,122 @@
+use compositor_api::types as api;
+use compositor_render::{
+    image::{ImageSource, ImageType},
+    InputId, OutputFrameFormat, OutputId, Renderer, RendererId, RendererSpec,
+};
+use input_uploader::InputUploader;
+use output_downloader::OutputDownloader;
+use types::to_js_error;
+use wasm_bindgen::prelude::*;
+
+mod app;
+mod input_uploader;
+mod output_downloader;
+mod types;
+mod utils;
+
+// Executed during WASM module init
+#[wasm_bindgen(start)]
+pub fn start() -> Result<(), JsValue> {
+    console_error_panic_hook::set_once();
+    tracing_wasm::set_as_global_default();
+    wasm_log::init(wasm_log::Config::new(log::Level::Trace));
+
+    Ok(())
+}
+
+#[wasm_bindgen(js_name = createRenderer)]
+pub async fn create_renderer(options: JsValue) -> Result<LiveCompositorRenderer, JsValue> {
+    let (device, queue, surface) = utils::create_wgpu_context().await?;
+    let mut options: compositor_render::RendererOptions =
+        types::from_js_value::<types::RendererOptions>(options)?.into();
+    options.wgpu_ctx = Some((device, queue));
+
+    let (renderer, _) = Renderer::new(options).map_err(to_js_error)?;
+    let input_uploader = InputUploader::default();
+    let output_downloader = OutputDownloader::default();
+
+    Ok(LiveCompositorRenderer {
+        renderer,
+        input_uploader,
+        output_downloader,
+        _surface: surface,
+    })
+}
+
+#[wasm_bindgen]
+pub struct LiveCompositorRenderer {
+    renderer: Renderer,
+    input_uploader: InputUploader,
+    output_downloader: OutputDownloader,
+    _surface: wgpu::Surface<'static>,
```

Comment:
What this `_surface` do?

Reply:
It represents a surface that you can render onto. In webGL you can't render in the headless mode and surface is required. AFAIK `wgpu` uses it to retrieve WebGL context